mirgecom/euler.py

```diff
@@ -96,11 +96,11 @@ class ConservedVars:  # FIXME: Name?
         Object array (:class:`~numpy.ndarray`) with shape ``(ndim,)``
         of :class:`~meshmode.dof_array.DOFArray`  for momentum density.
 
-    .. attribute:: mass_fractions
+    .. attribute:: scalar
```

Comment:
:shrug:

FWIW, here's my reasoning for suggesting `scalar`: [link](https://github.com/illinois-ceesd/mirgecom/pull/95#discussion_r558625893)

Reply:
I like scalars better, but it does not follow @majosm's reasoning. Since we don't have any other types of "scalars" - I named it mass_fractions (because those are the only types of scalars we have *outside of testing*).  Here's my reasoning:

The canonical conserved quantities for compressible fluids are:
Q = < mass density, momentum density, energy density, mass density fractions (for transported scalars)>

There are a couple of things to note about these:  
* From physics: The exactly conserved quantities that we care about are mass(in most contexts), momentum, energy, and charge
* For compressible fluids, the fluid state Q stores the conserved quantities over a test volume V, e.g. mass/V = mass *density*
* Mass fractions for transported scalars can include transported quantities that are not simple species (e.g. marker particles, charged species (ions), etc) - some of which participate in the mixture, and some don't at all or they participate in complicated ways as *defined by the EOS or mixture handler*)

My naming scheme was simple - since it is well known to anyone doing CFD that compressible codes work with conserved quantities per unit volume, I simply drop the *density* designation, so that the CV component names line up with the actual quantities that are conserved.  (And it's shorter to write!) That leads me to the CV I had (mass, momentum, energy, mass_fractions).

At risk of muddling this up with a different (but related) topic, another facet of this discussion is whether to impose any additional structure or requirements on "mass fractions". 

We should not enforce anything on the content of the *fractional mass densities* (rho*Y) of the transported scalars.  Whatever those are, the handling of them should be left to the construct that implements the mixture physics.  For example, Cantera (= pyrometheus) in our current code.  It is up to Cantera to handle how it treats the mass fractions, and whether they add to 1, or whether the inert/buffer species is explicitly tracked or not.  These considerations are controlled by the mixture physics package.  For plasmas, for example, Cantera tracks ions and other non-participatory species - which all still have mass fractions - and which all still must go into the Q/CV data structure, but imposing any external structure or requirement on those arrays or their content would be wrong.  In that case it is also not correct that SUM(Y) = 1.  There will be a subset of the Y for which that is true, but that subset is managed by Cantera, *not* the calling code.

Reply:
Hang on... I think I was confused in the other PR with what I said about the energy. I was thinking `E` was energy density, but it's not  -- `E` is _per unit mass_, not per unit volume. So `rho*E` is the energy density, and calling it `energy` is actually fine.

> The canonical conserved quantities for compressible fluids are:
Q = < mass density, momentum density, energy density, mass density fractions (for transported scalars)>

What does mass density fraction mean here?

Reply:
Actually, for the eqns we solve the cv.energy is this:

(1/V) * (mass * e +  kinetic_energy), with e = specific energy = (energy / unit mass)

The energy component of the CV is total energy density (i.e. [internal energy + kinetic energy] per unit volume).  I only intend to point out here that total energy E is not per unit mass.  Only (e) the specific internal energy is the per-unit-mass part.

Reply:
Now I'm even more confused... think it's time for a lunch break. 🙂

Reply:
> What does mass density fraction mean here?

The mass fraction (Y) = (mass of item) / (total mass)
The mass density fraction = (total mass/volume * Y) = (mass of item) / (volume)

Re: my previous argument about the naming scheme, I almost chose species_mass over mass_fraction, but I opted for mass_fraction because it is the accepted naming scheme we have for Y.  In some cases (as noted above) Y is not *actually* a mass fraction - but *we don't have any of those cases yet*!  Since all of our production cases have Y corresponding to species mass fractions - I committed to that and named it "mass_fractions" - even though I *test* scalar advection in a more general sense (i.e. as generic transported scalars).

Reply:
I think we were using different definitions of `E`. I haven't really looked at the DG book's Euler stuff, so I was going by the old CM/PC2 definition (which @anderson2981 tells me is the total energy per unit mass). By that definition `rho*E` is energy/volume (and is presumably what the book calls just `E`). In that sense calling it `ConservedVars.energy` makes sense, and my argument for using `scalar` instead of `scalars` is no longer valid. Either sounds good to me now. (Still opposed to `mass_fractions`, though; omitting a [m^-3] is different from omitting a multiplication by `rho`.)

Reply:
Independent of what reference or symbols are used, the "energy" term in the conservation form of the independent compressible fluid vars is "total energy density".  Some reference use "energy" some say "energy density", some say "total energy" - but they all have the same units: " energy / volume ".

Reply:
Anything you guys decide on naming this will be fine for me, especially among the current choices. 

My opinion remains that the naming should be consistent with the units of the respective components.  Since our vector of independent variables:
Q = [rho(mass density), rhoV(momentum density), rhoE(energy density), rhoY(fractional scalar mass density)], 
has units of [(mass/vol),(mom/vol),(energy/vol),(mass/vol)], that's where i got:
cv.{mass, momentum, energy, mass_fractions}.

The current suggestions are:
cv.{scalar, scalars, mass_fractions, scalar_mass}

I like any of these and think they are all consistent with the intended use.  However, for our purposes I maintain that we should retain the word "mass" in the name of the CV component for the fluid modules.  Then, in the future *if our needs dictate it* would we need any machinery to separate out scalar components that are *not* mixture species mass fractions.